packages/server/src/esignet-api.ts

```diff
@@ -0,0 +1,233 @@
+  code: z.string(),
+  clientId: z.string(),
+  redirectUri: z.string()
+});
+
+export type OIDPUserInfoRequest = FastifyRequest<{
+  Body: z.infer<typeof OIDPUserInfoSchema>;
+}>;
+
+type FetchTokenProps = {
+  code: string;
+  clientId: string;
+  redirectUri: string;
+  grantType?: string;
+};
+
+const generateSignedJwt = async (clientId: string) => {
+  const header = {
+    alg: JWT_ALG,
+    typ: "JWT",
+  };
+
+  const payload = {
+    iss: clientId,
+    sub: clientId,
+    aud: env.OIDP_JWT_AUD_CLAIM,
+  };
+
+  
+  const decodeKey = Buffer.from(
+    env.OIDP_CLIENT_PRIVATE_KEY!,
+    "base64"
+  )?.toString();
+  console.log("decodeKey: ", decodeKey)
+
+  const jwkObject = JSON.parse(decodeKey);
+  const privateKey = await jose.importJWK(jwkObject, JWT_ALG);
+
+  return new jose.SignJWT(payload)
+    .setProtectedHeader(header)
+    .setIssuedAt()
+    .setExpirationTime(JWT_EXPIRATION_TIME)
+    .sign(privateKey);
+};
+
+const fetchToken = async ({
+  code,
+  clientId,
+  redirectUri,
+}: FetchTokenProps) => {
+  const body = new URLSearchParams({
+    code: code,
+    client_id: clientId,
+    redirect_uri: redirectUri,
+    grant_type: "authorization_code",
+    client_assertion_type: "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+    client_assertion: await generateSignedJwt(clientId),
+  });
+
+  const request = await fetch(OIDP_TOKEN_ENDPOINT!, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/x-www-form-urlencoded",
+    },
+    body,
+  });
+  
+  const response = await request.json();
+  return response as { access_token?: string };
+};
+
+export const getOIDPUserInfo = async (
+  request: OIDPUserInfoRequest,
+  reply: FastifyReply
+) => {
+  const { code, clientId, redirectUri } = request.body;
+
+  const tokenResponse = await fetchToken({
+    code,
+    clientId,
+    redirectUri
+  });
+
+  if (!tokenResponse.access_token) {
+    throw new Error(
+      "Something went wrong with the OIDP token request. No access token was returned. Response from OIDP: " +
+        JSON.stringify(tokenResponse)
+    );
+  }
+
+  return fetchUserInfo(tokenResponse.access_token);
+};
+
+const decodeUserInfoResponse = (response: string) => {
+  return jwt.decode(response) as OIDPUserInfo;
+};
+
+export const fetchLocationFromFHIR = <T = any>(
+  suffix: string,
+  method = "GET",
+  body: string | undefined = undefined
+): Promise<T> => {
+  return fetch(`${env.GATEWAY_URL}${suffix}`, {
+    method,
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body,
+  })
+    .then((response) => {
+      return response.json();
+    })
+    .catch((error) => {
+      return Promise.reject(
+        new Error(`Fetch Location from FHIR request failed: ${error.message}`)
+      );
+    });
+};
+
+const searchLocationFromFHIR = (name: string) =>
+  fetchLocationFromFHIR<fhir2.Bundle>(
+    `/locations?${new URLSearchParams({ name, type: "ADMIN_STRUCTURE" })}`
+  );
```

Comment:
We need to add an issue to refactor this to use a GraphQL endpoint or something else than directory querying Hearth: but for first version to just get everything working it's fine. Locations... 🤔 🤔